File: BucketSort.java
import java.util.ArrayList;
import java.io.*;


public class BucketSort implements Runnable{
    int bucketCounter = 0;
    Bucket[] buckets;
    
    public BucketSort() {}
     
    public void run()
    {int myBucket;
        myBucket = takeBucket(); //Apenas uma thread por vez pode pegar um bucket para si
        insertionSort(buckets[myBucket].bucket); //Executar a ordenacao no bucket, entretanto, pode ser feito quando ela bem entender!
        System.out.println("Thread com bucket "+myBucket+" terminou a ordenacao!");
    }
    synchronized int takeBucket()
    {int myBucket;
        myBucket = bucketCounter;
        bucketCounter++;
        return myBucket;
    }
    public static void main(String args[])
    {long start,end;
        BucketSort bucketSort = new BucketSort();
        int qtdElementosVetor;
        int qtdThreads = 16;
        int[] vetor = null;

        int contElementosVetor = 0;
        
        try{
        // Open the file that is the first 
        // command line parameter
        FileInputStream fstream = new FileInputStream("randomInput.txt");
        // Get the object of DataInputStream
        DataInputStream in = new DataInputStream(fstream);
        BufferedReader br = new BufferedReader(new InputStreamReader(in));
        String strLine;
        //Read File Line By Line
        qtdElementosVetor = Integer.parseInt(br.readLine());
        
        //Gera vetor de numeros aleatorios
        vetor = new int[qtdElementosVetor];
        
        while ((strLine = br.readLine()) != null)   {
            vetor[contElementosVetor] = Integer.parseInt(strLine);
            contElementosVetor++;
        }
        //Close the input stream
        in.close();
          }catch (Exception e){//Catch exception if any
        System.err.println("Error: " + e.getMessage());
        }

        //Cria o vetor de buckets
        bucketSort.buckets = new Bucket[qtdThreads];  
        
        //Inicializa o vetor de buckets
        for(int i=0;i<qtdThreads;i++)
            bucketSort.buckets[i] = new Bucket();
        
        start = System.currentTimeMillis();
            
        //Criterio de separacao dos elementos por bucket de modo que os elementos 
        //em um bucket seja heterogeneo e possuam a mesma quantidade
        separaElementosDoVetorNosBuckets(vetor,bucketSort.buckets);
        
        end = System.currentTimeMillis();
        
        System.out.println("Tempo de execucao de alocacao foi: "+(end-start)+"ms.");
        
        //Inicia a criacao das threads para cuidar de cada bucket
        Thread[] threads = new Thread[qtdThreads];
        for(int i=0;i<qtdThreads;i++)
            threads[i] = new Thread(bucketSort);
        
        start = System.currentTimeMillis();
            
        //Inicie todas as threads para que elas apliquem InsertionSort em seus buckets
        for(int i=0;i<qtdThreads;i++)
            threads[i].start();
            
        //Espere que todas as threads terminem a ordenacao
        try {
            for(int i=0;i<qtdThreads;i++)
                threads[i].join();
            }catch (InterruptedException e){
                throw new RuntimeException(e);
            }
        
        end = System.currentTimeMillis();
        
        System.out.println("Tempo de execucao de ordenacao foi: "+(end-start)+"ms.");
        System.out.println();
        
            
        //Imprime os buckets em ordem crescente simulando a concatenacao 
        for(int i=0; i < bucketSort.buckets.length;i++)
        {
            System.out.println("Bucket "+i+", qtd elemns: {"+bucketSort.buckets[i].bucket.size()+"}:"+bucketSort.buckets[i]);
            System.out.println();
        }
    }
    private static void separaElementosDoVetorNosBuckets(int[] vetor, Bucket[]buckets)
    {
        for(int i = 0; i<vetor.length;i++)
            buckets[decidePosicaoDoBucket(vetor[i],buckets)].add(vetor[i]);
    }
    /**
        Este metodo dinamicamente atribui a cada bucket uma faixa de intervalo baseado
        na quantidade de elementos de cada bucket. 
        
        Ex: se vetor.size = 20 elementos, buckets.size = 4 buckets, as faixas de intervalo
        sao de bucket(1) = 1 a 5, bucket(2) = de 6 a 10, bucket(3) de 11 a 15 e bucket(4) 16 a 20.
    
        O criterio de alocao para numeros negativos e feito observando o modulo do valor. 
            Uma vez que a faixa de valores permitidas e simetrica (-5000 a +5000) a distribuicao
            dos elementos por bucket em teoria nao deve ser afetada.
            
        Se o elemento atual for 0 ele coloca imediatamente no primeiro bucket.
            
        Lembrar que o bucket sort CONCATENA os buckets ao final, logo os elementos de cada bucket
        devem necessariamente obedecer a uma faixa de intervalos. A heterogeneidade dos elementos
        de cada bucket deve obedecer, portanto, a esse intervalo.
    */
    private static int decidePosicaoDoBucket(int elementoAtual, Bucket[]buckets)
    {int i=0;
        if(elementoAtual == 0)
            return buckets.length/2;
                      
        int elementoAtualEmModulo = elementoAtual;
        
        if(elementoAtualEmModulo<0)
            elementoAtualEmModulo = elementoAtualEmModulo*-1;
            
        for(;i<(buckets.length/2);i++)
        {
            if(elementoAtualEmModulo < ((5000/((float)buckets.length/2))*(i+1)))
            {
                if(elementoAtual < 0)
                    return((buckets.length/2) - (i+1));
                else
                    return((buckets.length/2) + i);
            }
        }
        if(elementoAtual < 0)
            return (buckets.length/2 - i); 
        else
            return((buckets.length/2) + i-1); //So ocorre quando elemento igual a 5000
    }
    public static void insertionSort(ArrayList<Integer> a) {
        for (int i=1; i < a.size(); i++) {
            /* Insert a[i] into the sorted sublist */
	        int v = a.get(i);
	        int j;
	        for (j = i - 1; j >= 0; j--) {
	            if (a.get(j) <= v) break;
	            a.set(j+1,a.get(j));
	        }
	        a.set(j+1,v);
        }
    }
}

File: bucketSort.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

void splitElements(char ** unsorted);
int decidePosicaoDoBucket(int elementoAtual,int numFuncoes);
void separaElementosDoVetorNosBuckets(int *vetor,int bucket[300][16], int numProcessos, int posicao[], int numFuncoes);


int main(argc,argv) 
{
		//For file management
		FILE *fr;            /* declare the file pointer */
		
		//to use on for
		int i;



//		printf("I am the master process and my rank is %d!\n",myRank);
		
		//I need to start reading the functions from file, here I go..!
		
		
		if(!(fr=fopen("randomInput4.txt", "rt")))  /* open the file for reading */
		{
			printf("Erro! Impossivel abrir o arquivo!\n");
		  	exit(1);	
		}
	   	int numFuncoes,numProcessos=10;
	   	char numElemen[100];

		fgets(numElemen, sizeof(numElemen), fr);
		numFuncoes = atoi(numElemen);
		char line[numFuncoes][80];
		int unsort[numFuncoes];
		int bucket[300][16];
		int j=0;
	 	while( (fgets(line[j], sizeof(numElemen), fr) != NULL) && j < numFuncoes ) j++;
	
		fclose(fr);  /* close the file prior to exiting the routine */


//		for(i=0;i<numFuncoes;i++)
//			printf("Posicoes no vetor: %s\n",line[i]);
		for(i=0;i<numFuncoes;i++)
			unsort[i] = atoi(line[i]);
//		for(i=0;i<numFuncoes;i++)
//			printf("%d\n",unsort[i]);
		
		//para saber o num de elementos de cada bucket
		int numElemenBucket[numProcessos];
		separaElementosDoVetorNosBuckets(unsort,bucket,numProcessos,numElemenBucket,numFuncoes);

		
	for(i=0;i<numProcessos;i++)
	{
		printf("size: %d, bucket[%d] = ",numElemenBucket[i],i);
		for(j=0;j<numElemenBucket[i];j++)
			printf("%d|",bucket[i][j]);
		printf("\n");
	}
		
		
	return 0;
}
void separaElementosDoVetorNosBuckets(int *vetor,int bucket[300][16], int numProcessos, int posicao[16],int numFuncoes)
{int i,j;
	int posicaoBucket;
	
	for(i=0;i<numProcessos;i++)
		posicao[i] = 0;
	
    for(i = 0; i<numFuncoes;i++)
	{
		posicaoBucket = decidePosicaoDoBucket(vetor[i],numProcessos);
		bucket[posicaoBucket][posicao[posicaoBucket]] = vetor[i];
		posicao[posicaoBucket]++;
	}
	//for(i=0;i<numFuncoes;i++)
	//	printf("%d\n",posicao[i]);
/*		
	for(i=0;i<numProcessos;i++)
	{
		printf("size: %d, bucket[%d] = ",posicao[i],i);
		for(j=0;j<posicao[i];j++)
			printf("%d|",bucket[i][j]);
		printf("\n");
	}
*/

}
int decidePosicaoDoBucket(int elementoAtual,int numProcessos)
{
	int i=0;
	    if(elementoAtual == 0)
	        return numProcessos/2;

	    int elementoAtualEmModulo = elementoAtual;

	    if(elementoAtualEmModulo<0)
	        elementoAtualEmModulo = elementoAtualEmModulo*-1;

	    for(;i<(numProcessos/2);i++)
	    {
	        if(elementoAtualEmModulo < ((5000/((float)numProcessos/2))*(i+1)))
	        {
	            if(elementoAtual < 0)
	                return((numProcessos/2) - (i+1));
	            else
	                return((numProcessos/2) + i);
	        }
	    }
	    if(elementoAtual < 0)
	        return (numProcessos/2 - i); 
	    else
	        return((numProcessos/2) + i-1); //So ocorre quando elemento igual a 5000	
}